compiler/rustc_middle/src/mir/mono.rs

```diff
@@ -84,7 +84,14 @@ impl<'tcx> MonoItem<'tcx> {
             .debugging_opts
             .inline_in_all_cgus
             .unwrap_or_else(|| tcx.sess.opts.optimize != OptLevel::No)
-            && !tcx.sess.link_dead_code();
+            && !tcx.sess.link_dead_code()
+            && !tcx.sess.instrument_coverage();
```

Comment:
Wouldn't that break for stable feature `-C link-dead-code` as well, if that were the case?

Reply:
It sounds like you understand this better than I do. What I know is, this check is performed for `-C link-dead-code`, and the synthetic unused functions I'm creating are also dead code that we want to link. Without also adding this check for `-Z instrument-coverage`, some tests failed. Adding this check fixed the failed test, and the cause and effect seem logical since it's required for `link-dead-code`.

But I don't understand anything more than that. 

Do you have any ideas on how to write a test that proves your theory that this can fail under some conditions?

Reply:
I had to think some about the reason this is disabled for `link-dead-code`. Without this you would have multiple copies of the same function in the objects that get linked together. If `link-dead-code` is enabled, that means they would all get put in the final binary and you'd get a linker error.

I can imagine a similar thing happening for coverage, where we have multiple definitions of the same counters or functions (in particular it could be the synthetic unused functions as you say).

If that's the problem then this fix works. I have a feeling we might be able to tweak things to recover some inlining in the future.. through some special attributes on the counters or whatever is conflicting, or modeling new constraints into the cgu/monomorphization pipeline. If you agree, maybe a FIXME here would be helpful in finding it again.

> Do you have any ideas on how to write a test that proves your theory that this can fail under some conditions?

Try building the test that failed without this with `-Ccodegen-units=1`. If that passes then I think we're good.

Reply:
The test still passes with:

```shell
build/x86_64-unknown-linux-gnu/stage1/bin/rustc \
  src/test/run-make-fulldeps/coverage/closure.rs \
  '-Ccodegen-units=1' '-C opt-level=3' '-Zinstrument-coverage' --edition=2018 
```

(And with opt-level 2, 1, and 0/none)